roachtest: kv50/enc=false/nodes=4/cpu=96/batch=64 failed [SendError bubbles up]

roachtest.kv50/enc=false/nodes=4/cpu=96/batch=64 [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3359107&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3359107&tab=artifacts#/kv50/enc=false/nodes=4/cpu=96/batch=64) on master @ [ab1fc343c9a1140191f96353995258e609a84d02](https://github.com/cockroachdb/cockroach/commits/ab1fc343c9a1140191f96353995258e609a84d02):


```
		  | I210826 14:26:22.303165 1 workload/cli/run.go:367  [-] 1  DEPRECATION: the --init flag on "workload run" will no longer be supported after 19.2
		  | I210826 14:26:22.332804 1 workload/workloadsql/workloadsql.go:113  [-] 2  starting 1000 splits
		  | I210826 14:26:31.980697 643 workload/workloadsql/workloadsql.go:179  [-] 3  finished 1000 of 1000 splits
		  | I210826 14:26:31.981413 1 workload/cli/run.go:408  [-] 4  creating load generator...
		  | I210826 14:26:32.069848 1 workload/cli/run.go:439  [-] 5  creating load generator... done (took 88.434076ms)
		  | Error: ERROR: failed to send RPC: sending to all replicas failed; last error: routing information detected to be stale; lastErr: [NotLeaseHolderError] lease held by different store; r315: replica (n4,s4):3 not lease holder; current lease is repl=(n3,s3):1 seq=4 start=1629987984.434423504,48 epo=1 pro=1629987984.434423504,45 (SQLSTATE XXUUU)
		  | Error: COMMAND_PROBLEM: exit status 1
		  | (1) COMMAND_PROBLEM
		  | Wraps: (2) Node 5. Command with error:
		  |   | ``````
		  |   | ./workload run kv --init --histograms=perf/stats.json --concurrency=256 --splits=1000 --duration=10m0s --read-percent=50 --batch=64 {pgurl:1-4}
		  |   | ``````
		  | Wraps: (3) exit status 1
		  | Error types: (1) errors.Cmd (2) *hintdetail.withDetail (3) *exec.ExitError
		  |
		  | stdout:
		Wraps: (4) exit status 20
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *cluster.WithCommandDetails (4) *exec.ExitError

	monitor.go:128,kv.go:137,kv.go:246,test_runner.go:777: monitor failure: monitor task failed: t.Fatal() was called
		(1) attached stack trace
		  -- stack trace:
		  | main.(*monitorImpl).WaitE
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/monitor.go:116
		  | main.(*monitorImpl).Wait
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/monitor.go:124
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.registerKV.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/kv.go:137
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.registerKV.func3
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/kv.go:246
		  | main.(*testRunner).runTest.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:777
		Wraps: (2) monitor failure
		Wraps: (3) attached stack trace
		  -- stack trace:
		  | main.(*monitorImpl).wait.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/monitor.go:172
		Wraps: (4) monitor task failed
		Wraps: (5) attached stack trace
		  -- stack trace:
		  | main.init
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/monitor.go:81
		  | runtime.doInit
		  | 	/usr/local/go/src/runtime/proc.go:6309
		  | runtime.main
		  | 	/usr/local/go/src/runtime/proc.go:208
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1371
		Wraps: (6) t.Fatal() was called
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.withPrefix (5) *withstack.withStack (6) *errutil.leafError
```
<details><summary>Reproduce</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*kv50/enc=false/nodes=4/cpu=96/batch=64.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x28f1142]

goroutine 34401 [running]:
panic(0x4ab4b60, 0x8ebc610)
	/usr/local/go/src/runtime/panic.go:1065 +0x565 fp=0xc00b231c18 sp=0xc00b231b50 pc=0x48b265
runtime.panicmem()
	/usr/local/go/src/runtime/panic.go:212 +0x5b fp=0xc00b231c38 sp=0xc00b231c18 pc=0x4892bb
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:734 +0x173 fp=0xc00b231c70 sp=0xc00b231c38 pc=0x4a2fb3
github.com/cockroachdb/cockroach/pkg/sql/colexecop.Closers.CloseAndLogOnErr(0xc00b987c80, 0x1, 0x1, 0x69031d8, 0xc00c18e1b0, 0xc010a11500, 0x26)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecop/operator.go:157 +0xc2 fp=0xc00b231d08 sp=0xc00b231c70 pc=0x28f1142
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*ParallelUnorderedSynchronizer).init.func1.1(0xc00c18e1e0, 0xc003bfc300, 0x7f30d567e4f0, 0xc002de6340, 0xc00b987cc0, 0x1, 0x1, 0xc00b987c60, 0x1, 0x1, ...)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/parallel_unordered_synchronizer.go:227 +0x9e fp=0xc00b231d50 sp=0xc00b231d08 pc=0x2cddb9e
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*ParallelUnorderedSynchronizer).init.func1(0xc003bfc300, 0x7f30d567e4f0, 0xc002de6340, 0xc00b987cc0, 0x1, 0x1, 0xc00b987c60, 0x1, 0x1, 0xc00b987c80, ...)
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/parallel_unordered_synchronizer.go:278 +0xc19 fp=0xc00b231f78 sp=0xc00b231d50 pc=0x2cde8b9
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc00b231f80 sp=0xc00b231f78 pc=0x4c5cc1
created by github.com/cockroachdb/cockroach/pkg/sql/colexec.(*ParallelUnorderedSynchronizer).init
	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/parallel_unordered_synchronizer.go:216 +0x149
```

--- a/pkg/sql/colexec/BUILD.bazel
+++ b/pkg/sql/colexec/BUILD.bazel
@@ -71,7 +71,6 @@ go_library(
         "//pkg/sql/sqltelemetry",  # keep
         "//pkg/sql/types",
         "//pkg/util",
-        "//pkg/util/cancelchecker",
         "//pkg/util/duration",  # keep
         "//pkg/util/encoding",  # keep
         "//pkg/util/humanizeutil",

--- a/pkg/sql/colexec/parallel_unordered_synchronizer.go
+++ b/pkg/sql/colexec/parallel_unordered_synchronizer.go
@@ -22,8 +22,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/colexecop"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
-	"github.com/cockroachdb/cockroach/pkg/util"
-	"github.com/cockroachdb/cockroach/pkg/util/cancelchecker"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
 	"github.com/cockroachdb/errors"
 )
@@ -249,30 +247,16 @@ func (s *ParallelUnorderedSynchronizer) init() {
 				switch state {
 				case parallelUnorderedSynchronizerStateRunning:
 					if err := colexecerror.CatchVectorizedRuntimeError(s.nextBatch[inputIdx]); err != nil {
-						if s.Ctx.Err() == nil && s.cancelLocalInput[inputIdx] != nil {
-							if errors.Is(err, context.Canceled) || errors.Is(err, cancelchecker.QueryCanceledError) {
-								// The input context has been canceled, yet the
-								// main context of the synchronizer has not.
-								// This indicates that the synchronizer has
-								// transitioned into draining state and wanted
-								// this goroutine to stop whatever it was doing.
-								// Therefore, we swallow the error and proceed
-								// to draining.
-								//
-								// Note that we need the second part of the
-								// conditional in case the context cancellation
-								// was observed by the CancelChecker and was
-								// propagated as a query canceled error.
-								if util.CrdbTestBuild {
-									if s.getState() != parallelUnorderedSynchronizerStateDraining {
-										colexecerror.InternalError(errors.AssertionFailedf(
-											"unexpectedly the input context is canceled, the main " +
-												"context is not, and not in the draining state",
-										))
-									}
-								}
-								continue
-							}
+						if s.getState() == parallelUnorderedSynchronizerStateDraining && s.Ctx.Err() == nil && s.cancelLocalInput[inputIdx] != nil {
+							// The synchronizer has just transitioned into the
+							// draining state and eagerly canceled work of this
+							// input. That cancellation is likely to manifest
+							// itself as the context.Canceled error, but it
+							// could be another error too; in any case, we will
+							// swallow the error because the user of the
+							// synchronizer is only interested in the metadata
+							// at this point.
+							continue
 						}
 						sendErr(err)
 						return

